src/Lean/Server/Completion.lean

```diff
@@ -15,62 +15,141 @@ import Lean.Meta.Tactic.Apply
 import Lean.Meta.Match.MatcherInfo
 import Lean.Server.InfoUtils
 import Lean.Parser.Extension
+import Lean.Server.FileSource
+import Lean.Server.CompletionItemData
 
-namespace Lean.Server.Completion
-open Lsp
-open Elab
-open Meta
-open FuzzyMatching
-
-private partial def consumeImplicitPrefix (e : Expr) (k : Expr → MetaM α) : MetaM α := do
+private partial def Lean.Server.Completion.consumeImplicitPrefix (e : Expr) (k : Expr → MetaM α) : MetaM α := do
   match e with
   | Expr.forallE n d b c =>
     -- We do not consume instance implicit arguments because the user probably wants be aware of this dependency
     if c == .implicit then
-      withLocalDecl n c d fun arg =>
+      Meta.withLocalDecl n c d fun arg =>
         consumeImplicitPrefix (b.instantiate1 arg) k
     else
       k e
   | _ => k e
 
-private def isTypeApplicable (type : Expr) (expectedType? : Option Expr) : MetaM Bool :=
-  try
-    match expectedType? with
-    | none => return true
-    | some expectedType =>
-      let mut (numArgs, hasMVarHead) ← getExpectedNumArgsAux type
-      unless hasMVarHead do
-        let targetTypeNumArgs ← getExpectedNumArgs expectedType
-        numArgs := numArgs - targetTypeNumArgs
-      let (_, _, type) ← forallMetaTelescopeReducing type (some numArgs)
-      -- TODO take coercions into account
-      -- We use `withReducible` to make sure we don't spend too much time unfolding definitions
-      -- Alternative: use default and small number of heartbeats
-      withReducible <| withoutModifyingState <| isDefEq type expectedType
-  catch _ =>
-    return false
+namespace Lean.Lsp
+
+inductive CompletionIdentifier where
+  | const (declName : Name)
+  | fvar  (id : FVarId)
+
+instance : ToJson CompletionIdentifier where
+  toJson
+    | .const declName => Json.mkObj [("const", toJson declName)]
+    | .fvar id        => Json.mkObj [("fvar", toJson id)]
+
+instance : FromJson CompletionIdentifier where
+  fromJson? j := do
+    if let .ok declName := j.getObjValAs? Name "const" then
+      return .const declName
+    if let .ok id := j.getObjValAs? FVarId "fvar" then
+      return .fvar id
+    .error "invalid CompletionItemData"
```

Comment:
Don't we have `deriving` handlers for that?

Reply:
I genuinely wasn't aware that it works for inductives now.